### PR TITLE
feat: implement comprehensible metadata

### PR DESCRIPTION
**Pull request by Juiced66**
## What does this PR do ?
- Implement comprehensible metadatas (optional)
- Implement groups for metadatas (optional)
- Implement corresponding test
- Documentation

### Other changes

- Improve code documentation
- improve types

### CODE REVIEW

File: lib/modules/model/types/ModelContent.ts [hunk 1]
```diff
@@ -46,7 +46,41 @@ export interface AssetModelContent extends KDocumentContent {
...
+     *   "extTemp": {
+     *     "group": "buildingEnv",
+     *     "locales": {
+     *     "en": {
+     *        "friendlyName": "External temperature",
+     *        "description": "Building external temperature"
+     *     },
+     *     "fr": {
+     *       "friendlyName": "Température extérieure",
+     *       "description": "Température à l'exterieur du building"
```
Comment (rolljee) [inline lib/modules/model/types/ModelContent.ts:62]:
```suggestion
     *       "description": "Température à l'exterieur du bâtiment"
```

File: lib/modules/model/ModelsRegister.ts [hunk 2]
```diff
@@ -46,40 +46,85 @@ export class ModelsRegister {
...
   registerAsset(
     engineGroup: string,
     model: string,
     measures: NamedMeasures,
     metadataMappings: JSONObject = {},
     defaultMetadata: JSONObject = {},
+    metadataDetails: JSONObject = {},
+    metadataGroups: JSONObject = {},
   ) {
+    // Check if the model name is PascalCase
```
Comment (rolljee) [inline lib/modules/model/ModelsRegister.ts:70]: This comment is not relevant

File: lib/modules/model/types/ModelContent.ts [hunk 3]
```diff
@@ -94,7 +127,41 @@ export interface DeviceModelContent extends KDocumentContent {
...
+    /**
+     * Metadata details
+     * @example
+     * {
+     *   "trackerType": {
+     *     "group": "sensorSpecs",
+     *     "locales": {
+     *     "en": {
+     *        "friendlyName": "Tracker version",
+     *        "description": "Firmware version of the tracker"
```
Comment (rolljee) [inline lib/modules/model/types/ModelContent.ts:139]:
```suggestion
     *        "description": "Firmware version of the sensor"
```

File: lib/modules/model/types/ModelDefinition.ts [hunk 4]
```diff
@@ -18,6 +17,34 @@ import { Decoder, NamedMeasures } from "../../../modules/decoder";
...
+ *       "extTemp": {
+ *         "group": "buildingEnv",
+ *         "locales": {
+ *           "en": {
+ *             "groupFriendlyName": "Building environment",
+ *             "friendlyName": "External temperature",
+ *             "description": "Container external temperature"
+ *           },
+ *           "fr": {
+ *             "friendlyName": "Température externe",
```
Comment (rolljee) [inline lib/modules/model/types/ModelDefinition.ts:31]: French translation does not have a groupFriendlyName

File: lib/modules/plugin/DeviceManagerPlugin.ts [hunk 5]
```diff
@@ -124,6 +159,36 @@ export class DeviceManagerPlugin extends Plugin {
...
+       *     defaultMetadata: {
+       *       company: "Acme Inc"
+       *     },
+       *     metadataDetails: {
+       *       trackerType: {
+       *         group: "sensorSpecs",
+       *         locales: {
+       *           en: {
+       *             friendlyName: "Tracker version",
+       *             description: "Firmware version of the tracker"
```
Comment (rolljee) [inline lib/modules/plugin/DeviceManagerPlugin.ts:172]:
```suggestion
       *             description: "Firmware version of the sensor"
```

File: tests/application/assets/Container.ts [hunk 6]
```diff
@@ -48,12 +46,95 @@ export const containerAssetDefinition: AssetModelDefinition = {
...
+    extTemp: {
+      group: "environment",
+      locales: {
+        en: {
+          groupFriendlyName: "Environment",
+          friendlyName: "External Temperature",
+          description: "The temperature outside the container"
+        },
+        fr: {
+          friendlyName: "Température Externe",
```
Comment (rolljee) [inline tests/application/assets/Container.ts:59]: No groupFirnedlyName for the French translation ?

**Review by rolljee — CHANGES_REQUESTED** [review]

File: lib/modules/model/types/ModelContent.ts (hunk 1 above)
Comment (Juiced66) [reply lib/modules/model/types/ModelContent.ts]: :+1:

File: lib/modules/model/ModelsRegister.ts (hunk 2 above)
Comment (Juiced66) [reply lib/modules/model/ModelsRegister.ts:70]: removed

File: lib/modules/model/types/ModelContent.ts (hunk 3 above)
Comment (Juiced66) [reply lib/modules/model/types/ModelContent.ts:139]: Replaced tracker by sensor everywhere

File: lib/modules/model/types/ModelDefinition.ts (hunk 4 above)
Comment (Juiced66) [reply lib/modules/model/types/ModelDefinition.ts:30]: it's an error, groupFriendlyName is only present in metadataGroups object I removed all errors

File: lib/modules/plugin/DeviceManagerPlugin.ts (hunk 5 above)
Comment (Juiced66) [reply lib/modules/plugin/DeviceManagerPlugin.ts]: :+1:

File: tests/application/assets/Container.ts (hunk 6 above)
Comment (Juiced66) [reply tests/application/assets/Container.ts:59]: same as said earlier, it's an error, changed my mind when developping and not change every where sry

File: lib/modules/model/ModelsRegister.ts [hunk 7]
```diff
@@ -46,40 +46,83 @@ export class ModelsRegister {
...
+   * @throws PluginImplementationError if the model name is not in PascalCase.
+   */
   registerAsset(
     engineGroup: string,
     model: string,
     measures: NamedMeasures,
     metadataMappings: JSONObject = {},
     defaultMetadata: JSONObject = {},
+    metadataDetails: JSONObject = {},
+    metadataGroups: JSONObject = {},
```
Comment (Juiced66) [inline lib/modules/model/ModelsRegister.ts:68]: Should we type better those objects ?

File: lib/modules/model/ModelsRegister.ts [hunk 8]
```diff
@@ -46,40 +46,83 @@ export class ModelsRegister {
...
+   * @param metadataGroups - Optional groups for organizing metadata, with localizations.
+   * @throws PluginImplementationError if the model name is not in PascalCase.
+   */
   registerDevice(
     model: string,
     measures: NamedMeasures,
     metadataMappings: JSONObject = {},
     defaultMetadata: JSONObject = {},
+    metadataDetails: JSONObject = {},
+    metadataGroups: JSONObject = {},
```
Comment (Juiced66) [inline lib/modules/model/ModelsRegister.ts:108]: should we type better those objects ?

File: lib/modules/model/ModelsRegister.ts (hunk 7 above)
Comment (Juiced66) [reply lib/modules/model/ModelsRegister.ts]: I did a commit, I'll rollback if needed

**Review by rolljee — APPROVED** [review]
I find that the documentation lack a bit of use case, but the code lgtm

**Review by rolljee — APPROVED** [review]

File: doc/2/controllers/models/write-device/index.md [hunk 9]
```diff
@@ -50,7 +78,9 @@ Method: POST
 
 - `model`: Device model name
 - `metadataMappings`: Mappings of the metadata in Elasticsearch format
-- `defaultValues`: Default values for the metadata
+- `defaultValues`: Default values for the metadata- 
+- `metadataDetails`: Metadata group and translations (currently supporting 'en' and 'fr')
```
Comment (Juiced66) [inline doc/2/controllers/models/write-device/index.md:82]: This is an error, we support all languages with this PR

File: doc/2/controllers/models/write-device/index.md [hunk 10]
```diff
@@ -36,7 +36,35 @@ Method: POST
     },
     "defaultValues": {
       // Default values for metadata
-    }
+    },
+    "metadataDetails": {
+      /*
+        Metadata details including tanslations and group.
+          [name: string]: {
+            group: string;
```
Comment (Juiced66) [inline doc/2/controllers/models/write-device/index.md:44]: should group be an optionnal propertie ?

**Pull request merged**